pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -410,3 +416,200 @@ func enrichLegacyObject(originalObj, returnedObj runtime.Object) error {
 	accessorReturned.SetUID(accessorOriginal.GetUID())
 	return nil
 }
+
+func getSyncRequester(orgId int64) *identity.StaticRequester {
+	return &identity.StaticRequester{
+		Type:           claims.TypeServiceAccount, // system:apiserver
+		UserID:         1,
+		OrgID:          orgId,
+		Name:           "admin",
+		Login:          "admin",
+		OrgRole:        identity.RoleAdmin,
+		IsGrafanaAdmin: true,
+		Permissions: map[int64]map[string][]string{
+			orgId: {
+				"*": {"*"}, // all resources, all scopes
+			},
+		},
+	}
+}
+
+type syncItem struct {
+	name       string
+	objStorage runtime.Object
+	objLegacy  runtime.Object
+}
+
+func getList(ctx context.Context, obj rest.Lister, listOptions *metainternalversion.ListOptions) ([]runtime.Object, error) {
+	ll, err := obj.List(ctx, listOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	return meta.ExtractList(ll)
+}
+
+func mode2DataSyncer(ctx context.Context, legacy LegacyStorage, storage Storage, kind string, reg prometheus.Registerer, serverLockService ServerLockService, requestInfo *request.RequestInfo) (bool, error) {
+	metrics := &dualWriterMetrics{}
+	metrics.init(reg)
+
+	log := klog.NewKlogr().WithName("DualWriterMode2Syncer")
+
+	everythingSynced := false
+	outOfSync := 0
+	syncSuccess := 0
+	syncErr := 0
+
+	maxInterval := dataSyncerInterval + 5*time.Minute
+
+	// LockExecuteAndRelease ensures that just a single Grafana server acquires a lock at a time
+	// The parameter 'maxInterval' is a timeout safeguard, if the LastExecution in the
+	// database is older than maxInterval, we will assume the lock as timeouted. The 'maxInterval' parameter should be so long
+	// that is impossible for 2 processes to run at the same time.
+	err := serverLockService.LockExecuteAndRelease(ctx, "dualwriter mode 2 sync", maxInterval, func(context.Context) {
+		log.Info("starting dualwriter mode 2 sync")
+		startSync := time.Now()
+
+		orgId := int64(1)
+
+		ctx = klog.NewContext(ctx, log)
+		ctx = identity.WithRequester(ctx, getSyncRequester(orgId))
+		ctx = request.WithNamespace(ctx, requestInfo.Namespace)
+		ctx = request.WithRequestInfo(ctx, requestInfo)
+
+		legacyList, err := getList(ctx, legacy, &metainternalversion.ListOptions{})
+		if err != nil {
+			log.Error(err, "unable to extract list from legacy storage")
+			return
+		}
+		log.Info("got items from legacy storage", "items", len(legacyList))
+
+		storageList, err := getList(ctx, storage, &metainternalversion.ListOptions{})
+		if err != nil {
+			log.Error(err, "unable to extract list from storage")
+			return
+		}
+		log.Info("got items from unified storage", "items", len(storageList))
```

Comment:
We're going to need to think about how this scales out for large sets, we'll likely want to process them in chunks vs loading everything into memory.

Reply:
You are right. For playlists this might work just fine. However, for entities like dashboards, we are going to have performance issues for sure.

For this first iteration, in which we will be dealing with playlists, we have a couple of options:
- Limit the number of items returned by `getList` to, let's say, 100 or 1000 records, and trigger an error/alert if we get instances with more records than the threshold
- Use the internal KV storage to keep track of the last processed item. Not sure if `List` supports offsets. Also, not sure if all entities are going to support ordering.
- Do nothing and address this in a follow-up PR, so we unblock the migration of playlists to mode 3

@DanCech What do you think?

Reply:
1 and 3 sounds like a reasonable way to unblock this and move things forward.

Reply:
I have added a clause to abort the data syncer execution if unified storage returns more than 1000 records. 

I realized that as of today, LegacyStorage ignores completely the `Limit` option (at least for playlists), so right now we can't specify a batch size. We will need to address this in a follow up PR.

I hope that the rule of aborting the execution if we get more than 1000 records it's enough for now. What do you think?

https://github.com/grafana/grafana/blob/f01263803a6071749b20850f2c415578c58cb823/pkg/registry/apis/playlist/legacy_storage.go#L59-L75